packages/client-react-streaming/src/transportedQueryRef.ts

```diff
@@ -0,0 +1,116 @@
+ *
+ * A child client component reading the `TransportedQueryReference` via useReadQuery will suspend until the promise resolves.
+ */
+export interface TransportedQueryReference<
+  TData = unknown,
+  TVariables = unknown,
+> extends QueryReferenceBase<TData, TVariables> {
+  /**
+   * Only available in React Server Components.
+   * Will be `undefined` after being passed to Client Components.
+   *
+   * Returns a promise that resolves back to the `TransportedQueryReference` that can be awaited in RSC to suspend a subtree until the originating query has been loaded.
+   */
+  toPromise?: () => Promise<TransportedQueryReference>;
+}
+
+export interface InternalTransportedQueryRef<
+  TData = unknown,
+  TVariables = unknown,
+> extends TransportedQueryReference<TData, TVariables> {
+  __transportedQueryRef: true | QueryReference<any, any>;
+  options: TransportedQueryRefOptions;
+  queryKey: string;
+}
+
+export function createTransportedQueryRef<TData, TVariables>(
+  options: TransportedQueryRefOptions,
+  queryKey: string,
+  promise: Promise<any>
+): InternalTransportedQueryRef<TData, TVariables> {
+  const ref: InternalTransportedQueryRef<TData, TVariables> = {
+    __transportedQueryRef: true,
+    options,
+    queryKey,
+  };
+  Object.defineProperty(ref, "toPromise", {
+    value: () => promise.then(() => ref),
+    enumerable: false,
+  });
+  return ref;
+}
+
+export function reviveTransportedQueryRef(
+  queryRef: InternalTransportedQueryRef,
+  client: ApolloClient<any>
+): [QueryReference<any, any>, CacheKey] {
+  const hydratedOptions = deserializeOptions(queryRef.options);
+  const cacheKey: CacheKey = [
+    hydratedOptions.query,
+    canonicalStringify(hydratedOptions.variables),
+    queryRef.queryKey,
+  ];
+  if (queryRef.__transportedQueryRef === true) {
+    queryRef.__transportedQueryRef = wrapQueryRef(
+      getSuspenseCache(client).getQueryRef(cacheKey, () =>
+        client.watchQuery(hydratedOptions)
+      )
+    );
+  }
+  return [queryRef.__transportedQueryRef, cacheKey];
+}
+
+function isTransportedQueryRef(
+  queryRef: object
+): queryRef is InternalTransportedQueryRef {
+  return "__transportedQueryRef" in queryRef;
+}
+
+export function useWrapTransportedQueryRef<TData, TVariables>(
+  queryRef: QueryReference<TData, TVariables> | InternalTransportedQueryRef
+) {
+  const client = useApolloClient();
+  let cacheKey: CacheKey | undefined;
+  if (isTransportedQueryRef(queryRef)) {
+    [queryRef, cacheKey] = reviveTransportedQueryRef(queryRef, client);
+  }
+  const unwrapped = unwrapQueryRef<any>(queryRef);
+  useEffect(() => {
+    if (cacheKey) {
+      if (unwrapped.disposed) {
+        getSuspenseCache(client).add(cacheKey, unwrapped);
+      }
+    }
+    // Omitting the deps is intentional. This avoids stale closures and the
+    // conditional ensures we aren't running the logic on each render.
+  });
```

Comment:
@jerelmiller should I still do all this?

Reply:
I left it for `useBackgroundQuery` [here](https://github.com/apollographql/apollo-client/pull/11821/files#diff-61cffa353d7574ee991f11babdad1bca78cdac9b4a2ccc4b8d938abf0b28bc2bR247-R256) to avoid issues where the `queryRef` gets disposed by the auto dispose timeout and the hook re-renders (which was a bug before that change in 3.9.10). I did update that check to use a `setTimeout` though since `dispose` is now again wrapped in a `setTimeout`.

I don't remember exactly where you use this hook, but if you're in the same situation as what I described above, then keeping this makes sense.

Reply:
This wraps around useReadQuery and useQueryRefHandlers, and it can also create a queryRef by reviving a transportedQueryRef. We can reduce these effects once React deduplicates transported objects, but for now they're fixed.